tests/aws/services/stepfunctions/v2/test_sfn_api.py

```diff
@@ -315,6 +317,102 @@ def test_list_sms(
+            return filtered_sms
+
+        # expect all state machines to be present
+        wait_until(lambda: _list_state_machines(expected_results_count=13), max_retries=20)
+
+        paginator = aws_client.stepfunctions.get_paginator("list_state_machines")
+        page_iterator = paginator.paginate(maxResults=5)
+
+        # Paginates across all results and filters out any StateMachines not relevant to the test
+        def _verify_paginate_results() -> list:
+            filtered_state_machines = []
+            for page in page_iterator:
+                assert 0 < len(page["stateMachines"]) <= 5
+
+                filtered_page = [sm for sm in page["stateMachines"] if sm["name"] in sm_names]
+                if filtered_page:
+                    sm_name_set = {sm.get("name") for sm in filtered_state_machines}
+                    # assert that none of the State Machines being added are already present
+                    assert not any(sm.get("name") in sm_name_set for sm in filtered_page)
+
+                    filtered_state_machines.extend(filtered_page)
+
+            assert len(filtered_state_machines) == len(sm_names)
+            return filtered_state_machines
+
+        # Since ListStateMachines is eventually consistent, we should re-attempt pagination
+        listed_state_machines = retry(_verify_paginate_results, retries=20, sleep=1)
+        sfn_snapshot.match("list-state-machines-page-2", listed_state_machines[-3:])
+
+        # maxResults value is out of bounds
+        with pytest.raises(Exception) as err:
+            aws_client.stepfunctions.list_state_machines(maxResults=1001)
+        sfn_snapshot.match("list-state-machines-invalid-param-too-large", err.value.response)
+
+        # nextToken is too short
+        with pytest.raises(Exception) as err:
+            aws_client.stepfunctions.list_state_machines(nextToken="")
+        sfn_snapshot.match(
+            "list-state-machines-invalid-param-short-nextToken",
+            {"exception_typename": err.typename, "exception_value": err.value},
+        )
+
+        # nextToken is too long
+        invalid_long_token = "x" * 1025
+        with pytest.raises(Exception) as err:
+            aws_client.stepfunctions.list_state_machines(nextToken=invalid_long_token)
+        sfn_snapshot.add_transformer(
+            RegexTransformer(invalid_long_token, f"<invalid_token_{len(invalid_long_token)}_chars>")
+        )
+        sfn_snapshot.match("list-state-machines-invalid-param-long-nextToken", err.value.response)
+
+        # where maxResults is 0, the default of 100 is used
+        retry(
+            lambda: _list_state_machines(expected_results_count=13, maxResults=0),
+            retries=20,
+            sleep=1,
+        )
+
+        for state_machine_arn in state_machine_arns:
+            aws_client.stepfunctions.delete_state_machine(stateMachineArn=state_machine_arn)
+
+        # expect no state machines to be leftover after deletion
```

Comment:
nit: more precisely, no filtered state machines that were created in this test :)